rust/wasmtime_bindings/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+    if let Err(e) = wasmtime_wasi::tokio::add_to_linker(&mut linker, |s| s) {
+        return Err(anyhow!("Failed to add wasi to linker: {:?}", e));
+    }
+    let mut inst_fut =
+        Box::pin(linker.instantiate_async(&mut store.wasmtime_store, &module.wasmtime_module));
+    let waker = futures::task::noop_waker_ref();
+    let mut ctx = Context::from_waker(&waker);
+
+    loop {
+        match inst_fut.as_mut().poll(&mut ctx) {
+            Poll::<Result<wasmtime::Instance>>::Pending => {}
+            Poll::<Result<wasmtime::Instance>>::Ready(Ok(inst)) => {
+                return Ok(Box::new(Instance {
+                    wasmtime_instance: inst,
+                }));
+            }
+            Poll::<Result<wasmtime::Instance>>::Ready(Err(e)) => {
+                return Err(anyhow!("Failed to instantiate module: {:?}", e));
+            }
+        }
+    }
+}
+
+pub struct Module {
+    wasmtime_module: wasmtime::Module,
+}
+
+fn create_module(engine: &mut Engine, script: &str) -> Result<Box<Module>> {
+    match wasmtime::Module::new(&engine.wasmtime_engine, script) {
+        Ok(module) => Ok(Box::new(Module {
+            wasmtime_module: module,
+        })),
+        Err(e) => Err(anyhow!("Compilation failed: {}", e)),
+    }
+}
+
+pub struct Store {
+    wasmtime_store: wasmtime::Store<wasmtime_wasi::WasiCtx>,
+}
+
+fn create_store(engine: &mut Engine) -> Result<Box<Store>> {
+    let wasi = wasmtime_wasi::tokio::WasiCtxBuilder::new().build();
+    let mut store = wasmtime::Store::new(&mut engine.wasmtime_engine, wasi);
+    // TODO: make this configurable
+    store.out_of_fuel_async_yield(10000, 4096);
+    Ok(Box::new(Store {
+        wasmtime_store: store,
+    }))
+}
+
+pub struct Memory {
+    wasmtime_memory: wasmtime::Memory,
+}
+
+impl Memory {
+    fn data(self: &Memory, store: &Store) -> *mut u8 {
+        self.wasmtime_memory.data_ptr(&store.wasmtime_store)
+    }
+    fn size(self: &Memory, store: &Store) -> u64 {
+        self.wasmtime_memory.size(&store.wasmtime_store)
+    }
+    fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64> {
+        self.wasmtime_memory.grow(&mut store.wasmtime_store, delta)
+    }
+}
+
+fn get_memory(instance: &Instance, store: &mut Store) -> Result<Box<Memory>> {
+    match instance
+        .wasmtime_instance
+        .get_export(&mut store.wasmtime_store, "memory")
+    {
+        None => Err(anyhow!(
+            "Memory export not found - please export `memory` in the wasm module"
+        )),
+        Some(export) => match export.into_memory() {
+            None => Err(anyhow!(
+                "Exported object memory is not a WebAssembly memory"
+            )),
+            Some(memory) => Ok(Box::new(Memory {
+                wasmtime_memory: memory,
+            })),
+        },
+    }
+}
+
+fn get_abi(instance: &Instance, store: &mut Store, memory: &Memory) -> Result<u32> {
+    match instance
```

Comment:
Style suggestion: the function returns a `Result<>`, so you can use the `?` operator and some methods on `Option` to make the code flatter. For example:

```rust
let export = instance
    .wasmtime_instance
    .get_export(&mut store.wasmtime_store, "_scylla_abi")
    .ok_or_else(|| anyhow!("ABI version export not found - please export `_scylla_abi` in the wasm module"))?;

let global = global
    .get(&mut store.wasmtime_store)
    .ok_or_else(|| anyhow!("Exported object _scylla_abi is not a global"))?;

// ...
```

Other functions would benefit from this conversion, too.

Reply:
I also added some `map_err` in the same style in the rebase